File: concurrency-and-node/haskell-async/src/LocalMain.hs
{-# LANGUAGE BangPatterns #-}

import Control.Concurrent.Async
import Data.Time

main :: IO ()
main = do
  testWithoutAsync
  testWithAsync

testWithoutAsync :: IO ()
testWithoutAsync = do
  !t1 <- getCurrentTime
  let !x = getFibs 43
      !y = getFibs 44
      !sum = x + y
  !t2 <- getCurrentTime
  putStrLn $ "Without Async"
  putStrLn $ "Sum: " ++ show sum
  putStrLn $ "Execution Time: " ++ show (diffUTCTime t2 t1)
{-# NOINLINE testWithoutAsync #-}

testWithAsync :: IO ()
testWithAsync = do
  !t1 <- getCurrentTime
  (x, y) <- concurrently
            (do let !f = getFibs 43
                return f)
            (do let !f = getFibs 44
                return f)
  let !sum = x + y
  !t2 <- getCurrentTime
  putStrLn $ "With Async"
  putStrLn $ "Sum: " ++ show sum
  putStrLn $ "Execution Time: " ++ show (diffUTCTime t2 t1)
{-# NOINLINE testWithAsync #-}

getFibs :: Int -> Int
getFibs 1 = 1
getFibs 2 = 1
getFibs n
  | n < 1     = error "Fib number is less than 1."
  | otherwise = getFibs (n - 1) + getFibs (n - 2)

File: concurrency-and-node/node-async/src/local-main.js
async = require('promise-async');

function main() {
    testWithoutAsync()
    testWithAsync();
}

function testWithoutAsync() {
    var x;
    var y;
    var t1 = new Date().getTime();

    x = getFibs(43);
    y = getFibs(44);

    var sum = x + y
    var t2 = new Date().getTime();

    console.log('Without Async');
    console.log('Sum: ' + sum);
    console.log('Execution Time: ' + ((t2 - t1) / 1000) + 's');
}

function testWithAsync() {
    var x;
    var y;
    var t1 = new Date().getTime();

    return async.parallel([
        (callback) => {
            x = getFibs(43);
            callback();
        },
        (callback) => {
            y = getFibs(44);
            callback();
        }
    ]).then(() => {
        var sum = x + y;
        var t2 = new Date().getTime();

        console.log('With Async');
        console.log('Sum: ' + sum);
        console.log('Execution Time: ' + ((t2 - t1) / 1000) + 's');
    });
}

function getFibs(n) {
    if (n < 1) {
        throw 'Fib number is less than 1.';
    } else if (n == 1) {
        return 1;
    } else if (n == 2) {
        return 1;
    } else {
        return getFibs(n - 1) + getFibs(n - 2);
    }
}

main();
